File: qif4j/src/test/java/com/github/squirrelgrip/qif4j/write/QifWriterTest.java
package com.github.squirrelgrip.qif4j.write;

import com.github.squirrelgrip.qif4j.QifAccount;
import com.github.squirrelgrip.qif4j.QifInterestIncome;
import com.github.squirrelgrip.qif4j.QifInvestment;
import com.github.squirrelgrip.qif4j.QifTransaction;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class QifWriterTest {

	@Test
	public void testEmptyAccountRegister() throws IOException {
		Writer writer = new StringWriter();
		QifWriter qifWriter = new QifWriter(writer);
		List<QifTransaction> txnList = new ArrayList<QifTransaction>();
		QifAccount account = TestUtils.buildPortfolioAccount();
		qifWriter.write(account, txnList);
		String expected = TestUtils
				.readTestSampleFile("src/test/resources/writer/invst_account_empty.qif");
		assertEquals(expected, writer.toString());
	}

	@Test
	public void testAccountRegisterWithSingleTxn() throws IOException {
		Writer writer = new StringWriter();
		QifWriter qifWriter = new QifWriter(writer);
		List<QifTransaction> txnList = new ArrayList<QifTransaction>();
		QifInvestment invst = TestUtils.buildInvst(10.952d, 100d, 9.95d);
		txnList.add(invst);
		QifAccount account = new QifAccount(QifAccountTypeEnum.PORTFOLIO,
				"Foo Portfolio", "My description", BigDecimal.valueOf(1000.07d));
		qifWriter.write(account, txnList);

		String expected = TestUtils
				.readTestSampleFile("src/test/resources/writer/invst_account_single.qif");
		assertEquals(expected, writer.toString());
	}

	@Test
	public void testAccountRegisterWithMultipleTxn() throws IOException {
		Writer writer = new StringWriter();
		QifWriter qifWriter = new QifWriter(writer);
		List<QifTransaction> txnList = new ArrayList<QifTransaction>();
		QifInvestment invst = TestUtils.buildInvst(10.952d, 100d, 9.95d);
		invst.setMemo("Buy 100 FOOB for $10.952. Commission: $9.95");
		txnList.add(invst);

		invst = TestUtils.buildInvst(13.05d, 100d, 9.95d);
		invst.setMemo("Sell 100 FOOB for capital gain. Commission: $9.95");
		invst.setAction(QifInvestmentAction.SELL.getText());
		invst.setDate(LocalDate.of(2014, 4, 5));
		txnList.add(invst);

		QifInterestIncome ii = new QifInterestIncome(
				QifAccountTypeEnum.PORTFOLIO, LocalDate.of(2014, 4, 6),
				BigDecimal.valueOf(0.12d));
		txnList.add(ii);

		invst = TestUtils.buildDividend(12.99d, 7.5d, ".15 on 50 shares");
		invst.setDate(LocalDate.of(2014, 5, 30));
		txnList.add(invst);

		invst = new QifInvestment();
		invst.setDate(LocalDate.of(2014, 6, 13));
		invst.setAction(QifInvestmentAction.TRANSFER_IN.getText());
		invst.setTotal(BigDecimal.valueOf(500d));
		invst.setCategory("[Test Checking]");
		txnList.add(invst);

		QifAccount account = new QifAccount(QifAccountTypeEnum.PORTFOLIO,
				"Foo Portfolio", "My description", BigDecimal.valueOf(1000.07d));
		qifWriter.write(account, txnList);

		String expected = TestUtils
				.readTestSampleFile("src/test/resources/writer/invst_account_multi.qif");
		assertEquals(expected, writer.toString());
	}
}

File: qif4k/src/test/kotlin/com/github/squirrelgrip/qif4k/write/QifWriterTest.kt
package com.github.squirrelgrip.qif4k.write

import com.github.squirrelgrip.qif4k.QifAccount
import com.github.squirrelgrip.qif4k.QifInterestIncome
import com.github.squirrelgrip.qif4k.QifInvestment
import com.github.squirrelgrip.qif4k.QifTransaction
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.io.StringWriter
import java.io.Writer
import java.math.BigDecimal
import java.time.LocalDate

class QifWriterTest {
    @Test
    fun testEmptyAccountRegister() {
        val writer: Writer = StringWriter()
        val qifWriter = QifWriter(writer)
        val txnList: List<QifTransaction> = mutableListOf()
        val account = TestUtils.buildPortfolioAccount()
        qifWriter.write(account, txnList)
        val expected = TestUtils.readTestSampleFile("src/test/resources/writer/invst_account_empty.qif")
        Assertions.assertEquals(expected, writer.toString())
    }

    @Test
    fun testAccountRegisterWithSingleTxn() {
        val writer: Writer = StringWriter()
        val qifWriter = QifWriter(writer)
        val txnList: MutableList<QifTransaction> = mutableListOf()
        val invst = TestUtils.buildInvst(10.952, 100.0, 9.95)
        txnList.add(invst)
        val account = QifAccount(QifAccountTypeEnum.PORTFOLIO,
                "Foo Portfolio", "My description", BigDecimal.valueOf(1000.07))
        qifWriter.write(account, txnList)
        val expected = TestUtils.readTestSampleFile("src/test/resources/writer/invst_account_single.qif")
        Assertions.assertEquals(expected, writer.toString())
    }

    @Test
    fun testAccountRegisterWithMultipleTxn() {
        val writer: Writer = StringWriter()
        val qifWriter = QifWriter(writer)
        val txnList: MutableList<QifTransaction> = mutableListOf()
        var invst = TestUtils.buildInvst(10.952, 100.0, 9.95)
        invst.memo = "Buy 100 FOOB for $10.952. Commission: $9.95"
        txnList.add(invst)
        invst = TestUtils.buildInvst(13.05, 100.0, 9.95)
        invst.memo = "Sell 100 FOOB for capital gain. Commission: $9.95"
        invst.action = QifInvestmentAction.SELL.text
        invst.date = LocalDate.of(2014, 4, 5)
        txnList.add(invst)
        val ii = QifInterestIncome(
                QifAccountTypeEnum.PORTFOLIO, LocalDate.of(2014, 4, 6),
                BigDecimal.valueOf(0.12))
        txnList.add(ii)
        invst = TestUtils.buildDividend(12.99, 7.5, ".15 on 50 shares")
        invst.date = LocalDate.of(2014, 5, 30)
        txnList.add(invst)
        invst = QifInvestment()
        invst.date = LocalDate.of(2014, 6, 13)
        invst.action = QifInvestmentAction.TRANSFER_IN.text
        invst.total = BigDecimal.valueOf(500.0)
        invst.category = "[Test Checking]"
        txnList.add(invst)
        val account = QifAccount(QifAccountTypeEnum.PORTFOLIO,
                "Foo Portfolio", "My description", BigDecimal.valueOf(1000.07))
        qifWriter.write(account, txnList)
        val expected = TestUtils.readTestSampleFile("src/test/resources/writer/invst_account_multi.qif")
        Assertions.assertEquals(expected, writer.toString())
    }
}